eval/expression.ts

```diff
@@ -0,0 +1,119 @@
+ *
+ * Note that although it inherits from primitive `string` for convenience, it
+ * actually inherits from the `String` wrapper object.
+ *
+ * ```typescript
+ * import { assertEquals } from "jsr:@std/assert/equals";
+ * import { expr } from "jsr:@denops/std/eval";
+ *
+ * const s: string = expr`foo`;
+ * assertEquals(typeof s, "object"); // is not "string"
+ * ```
+ */
+export type Expression = string & ExpressionProps;
+
+interface ExpressionProps extends VimEvaluatable {
+  /**
+   * Used by the `JSON.stringify` to enable the transformation of an object's data to JSON serialization.
+   */
+  toJSON(): string;
+
+  readonly [Symbol.toStringTag]: "Expression";
+}
+
+/**
+ * Create a {@linkcode Expression} that evaluates as a Vim expression.
+ *
+ * `raw_vim_expression` is a string text representing a raw Vim expression.
+ * Backslashes are treated as Vim syntax rather than JavaScript string escape
+ * sequences. Note that the raw Vim expression has not been verified and is
+ * therefore **UNSAFE**.
+ *
+ * `${value}` is the expression to be inserted at the current position, whose
+ * value is converted to the corresponding Vim expression. The conversion is
+ * safe and an error is thrown if it cannot be converted. Note that if the
+ * value contains `Expression` it will be inserted as-is, this is **UNSAFE**.
+ *
+ * ```typescript
+ * import { assertEquals } from "jsr:@std/assert/equals";
+ * import { expr } from "jsr:@denops/std/eval";
+ *
+ * assertEquals(
+ *   expr`raw_vim_expression`.toString(),
+ *   "raw_vim_expression",
+ * );
+ *
+ * const value = ["foo", 123, null];
+ * assertEquals(
+ *   expr`raw_vim_expression + ${value}`.toString(),
+ *   "raw_vim_expression + ['foo',123,v:null]"
+ * );
+ * ```
+ */
+export function expr(
+  template: TemplateStringsArray,
+  ...substitutions: TemplateSubstitutions
+): Expression {
+  const values = substitutions.map(stringify);
+  const raw = String.raw(template, ...values);
+  return new ExpressionImpl(raw) as unknown as Expression;
+}
+
+/**
+ * Returns `true` if the value is a {@linkcode Expression}.
+ *
+ * ```typescript
+ * import { assert, assertFalse } from "jsr:@std/assert";
+ * import { isExpression, expr } from "jsr:@denops/std/eval";
+ *
+ * assert(isExpression(expr`123`));
+ *
+ * assertFalse(isExpression("456"));
+ * ```
+ */
+export const isExpression: Predicate<Expression> = isIntersectionOf([
+  // NOTE: Do not check `isString` here, because `Expression` has a different type in definition (primitive `string`) and implementation (`String`).
+  isObjectOf({
+    [Symbol.toStringTag]: isLiteralOf("Expression"),
+  }),
+  isVimEvaluatable,
+]) as unknown as Predicate<Expression>;
+
+// deno-lint-ignore no-explicit-any
+type TemplateSubstitutions = any[];
+
+class ExpressionImpl extends String implements ExpressionProps {
+  declare [Symbol.toStringTag]: "Expression";
+  static {
+    this.prototype[Symbol.toStringTag] = "Expression";
+  }
```

Comment:
Why do we need to use `static` with `prototype` here? Can't we write it like

```ts
class ExpressionImpl extends String implements ExpressionProps {
  readonly [Symbol.toStringTag] = "Expression";
  // ...
}
```

Or

```ts
class ExpressionImpl extends String implements ExpressionProps {
  ger [Symbol.toStringTag]() {
    return "Expression";
  }
  // ...
}
```

Reply:
This is the same as defining the `name` property as a static property when inheriting from the Error class, because the value is a constant. This implementation is a bit more efficient, although as you say, it is a bit less readable.
The first example defines it as an instance property, but it doesn't have to be defined on each instance.
The second example defines it as a getter, but it doesn't have to be executed every time.

Reply:
There’s almost no significant difference among these three implementations, so I prioritize readability.

```
Benchmark       Time (avg)        Iter/s              (min … max)       p75       p99      p995
--------------------------------------------------------------- -----------------------------

Group: Constructor
Static          6.18 ms/iter         161.9     (6.04 ms … 6.33 ms) 6.28 ms 6.33 ms 6.33 ms
Attribute       6.29 ms/iter         158.9     (6.19 ms … 6.56 ms) 6.31 ms 6.56 ms 6.56 ms
Getter          6.08 ms/iter         164.5        (6 ms … 6.27 ms) 6.09 ms 6.27 ms 6.27 ms

Summary
  Static
   1.02x slower than Getter
   1.02x faster than Attribute

Group: toStringTag
Static         47.03 ms/iter          21.3   (46.67 ms … 48.07 ms) 47.06 ms 48.07 ms 48.07 ms
Attribute      46.95 ms/iter          21.3   (45.96 ms … 48.22 ms) 47.54 ms 48.22 ms 48.22 ms
Getter         55.74 ms/iter          17.9   (55.46 ms … 57.35 ms) 55.67 ms 57.35 ms 57.35 ms

Summary
  Static
   1x slower than Attribute
   1.19x faster than Getter
```

Given these results, the `readonly [Symbol.toStringTag] = "Expression";` approach seems to offer the best balance between performance and readability.

<details>
<summary>Code</summary>

```ts
class Static {
  declare [Symbol.toStringTag]: "A";
  static {
    this.prototype[Symbol.toStringTag] = "A";
  }
}

class Attribute {
  readonly [Symbol.toStringTag] = "B";
}

class Getter {
  get [Symbol.toStringTag]() {
    return "D";
  }
}

// Warm-up
Array.from({ length: 1000 }, () => {
  new Static();
  new Attribute();
  new Getter();
});

const count = 1_000_000;
const array = Array.from({ length: count });

Deno.bench({
  name: "Static",
  fn: () => array.forEach(() => new Static()),
  group: "Constructor",
  baseline: true,
});

Deno.bench({
  name: "Attribute",
  fn: () => array.forEach(() => new Attribute()),
  group: "Constructor",
});

Deno.bench({
  name: "Getter",
  fn: () => array.forEach(() => new Getter()),
  group: "Constructor",
});

Deno.bench({
  name: "Static",
  fn: () => {
    const toString = Object.prototype.toString.bind(new Static());
    array.forEach(() => toString());
  },
  group: "toStringTag",
  baseline: true,
});

Deno.bench({
  name: "Attribute",
  fn: () => {
    const toString = Object.prototype.toString.bind(new Attribute());
    array.forEach(() => toString());
  },
  group: "toStringTag",
});

Deno.bench({
  name: "Getter",
  fn: () => {
    const toString = Object.prototype.toString.bind(new Getter());
    array.forEach(() => toString());
  },
  group: "toStringTag",
});
```

</details>

Reply:
I agree with "prioritize readability" and would like to adopt it, but when I said "bit efficient" I meant memory usage, not speed.

Reply:
In that case, I believe readability is more important, unless the difference in memory usage is significant.

Reply:
I decided to make it a getter because I think should not add an own-property to an instance.